### PR TITLE
Yasmin- Leaves

### PR DESCRIPTION
**Pull request by YasminM11**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I used a custom method to get the top ten work. I also used .sort (spotlight).
Describe how you approached testing that model method. What edge cases did you come up with? | I tested the top ten.
What are `session` and `flash`? What is the difference between them? | session: to track the data for the user. Flash: is used to display a message to the user.
What was one thing that you gained more clarity on through this assignment? | how to use rails application and heroku app.
What is the Heroku URL of your deployed application? | https://media-ranker-yasmin.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | more tests practice.

### CODE REVIEW

File: test/models/vote_test.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    it "don't allow to vote more than once for the same work" do
+      vote1 = Vote.create(work: works(:work1), user: users(:user1))
+      vote1.save
+      vote2 = Vote.new(work: works(:work1), user: users(:user1))
+      result = vote2.valid?
+      expect(result).must_equal false
+    end
+
+    it "needs a username(id)" do
+      vote = Vote.new(work_id: @work)
```
Comment (beccaelenzil) [inline test/models/vote_test.rb:28]: This test is failing. I'm not entirely clear why, but one thing to note: you have a `@work` for this test. If you did, it should either be `work_id: @work.id` or `work: @work`

File: app/views/homepages/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      <span class="spotlight__header--prefix">Media Spotlight</span>
+      <%= link_to Work.spotlight.title, work_path(Work.spotlight.id), class: "spotlight__link-to" %> by <%= Work.spotlight.creator%>
+    </h2>
+    <p class="spotlight__description"><%=Work.spotlight.votes.count%> votes - <%=Work.spotlight.description%>
+    </p>
+  </section>
+
+<hr class="root__hr">
+
+<section class="top-ten__container">
```
Comment (beccaelenzil) [inline app/views/homepages/index.html.erb:14]: This is a great place for a partial view since you essentially repeat the same code 3 times.

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
+class Work < ApplicationRecord
+  has_many :votes
+  # validations
+  validates :title, presence: true, uniqueness: true  
+  
+  def self.top_ten(category)
```
Comment (beccaelenzil) [inline app/models/work.rb:6]: It would be good if this method also sorted the top ten by vote.

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+
+class UsersController < ApplicationController
+  # app/controllers/users_controller.rb
+  def index 
+    @users = User.all
+  end
+
+
+  def show
+    user_id = params[:id]
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:10]: Finding a logged in user is good functionality to put in a controller filter because it may be used multiple places in the future.

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  
+  # before_action :find_book, only: [:show, :edit, :update, :destroy]
+  # before_action :if _work_missing, only[:show, :edit, :destroy]
+  def index
+    @works = Work.all.order(:title) 
+    @all_movies = Work.where(category: "movie")
+    @all_books = Work.where(category: "book")
+    @all_albums = Work.where(category: "album")
+    #  @works = Work.all
+    @top_movies = Work.top_ten("movie")
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:11]: I think these are only needed in the homepages controller

File: test/models/work_test.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  describe 'top ten' do
+    it "return the top ten for each category" do
+      
+      top = Work.spotlight
+      all_work = Work.all
+      all_work.each do |work|
+        expect(top.votes.count).must_be :>=, work.votes.count
+      end
+    end
+  end
```
Comment (beccaelenzil) [inline test/models/work_test.rb:53]: remember to include a test for your spotlight method.

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes -- but not sorted by vote
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes -- but no flash message
Can go to "View all media" page and see three lists of works, sorted by vote | These are not sorted by vote
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | This produces a 404 page, a redirect would be better
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes -- consider changing the css so the flash message is more visible.
**Log out** |
Logging out showed a flash message and changed the UI |yes 
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | limit these routes to the ones your need
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | see comment

## Overall Feedback


Good work overall. Your implementation matches the demo site very closely and you did a good job testing the validations and relationships of models. I've left some inline comment for you to review. A couple things to note: you missed the requirement to sort the works by vote count and it would be good to make more use of partial views to DRY up your code. Keep up the hard work!

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes -- but not sorted by vote
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes -- but no flash message
Can go to "View all media" page and see three lists of works, sorted by vote | These are not sorted by vote
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | This produces a 404 page, a redirect would be better
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes -- consider changing the css so the flash message is more visible.
**Log out** |
Logging out showed a flash message and changed the UI |yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | limit these routes to the ones your need
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | see comment

## Overall Feedback


Good work overall. Your implementation matches the demo site very closely and you did a good job testing the validations and relationships of models. I've left some inline comment for you to review. A couple things to note: you missed the requirement to sort the works by vote count and it would be good to make more use of partial views to DRY up your code. Keep up the hard work!
